src/repr/src/row.rs

```diff
@@ -746,16 +874,58 @@ where
         Datum::False => data.push(Tag::False.into()),
         Datum::True => data.push(Tag::True.into()),
         Datum::Int16(i) => {
-            data.push(Tag::Int16.into());
-            data.extend_from_slice(&i.to_le_bytes());
+            let mbs = min_bits_signed(i);
+            if mbs <= 8 {
+                data.push(Tag::Int16_8.into());
+                data.extend_from_slice(&i.to_le_bytes()[0..1]);
+            } else {
+                data.push(Tag::Int16.into());
+                data.extend_from_slice(&i.to_le_bytes());
+            }
         }
         Datum::Int32(i) => {
-            data.push(Tag::Int32.into());
-            data.extend_from_slice(&i.to_le_bytes());
+            let mbs = min_bits_signed(i);
+            if mbs <= 8 {
+                data.push(Tag::Int32_8.into());
+                data.extend_from_slice(&i.to_le_bytes()[0..1]);
+            } else if mbs <= 16 {
+                data.push(Tag::Int32_16.into());
+                data.extend_from_slice(&i.to_le_bytes()[0..2]);
+            } else if mbs <= 24 {
+                data.push(Tag::Int32_24.into());
+                data.extend_from_slice(&i.to_le_bytes()[0..3]);
+            } else {
+                data.push(Tag::Int32.into());
+                data.extend_from_slice(&i.to_le_bytes());
+            }
         }
         Datum::Int64(i) => {
-            data.push(Tag::Int64.into());
-            data.extend_from_slice(&i.to_le_bytes());
+            let mbs = min_bits_signed(i);
+            if mbs <= 8 {
+                data.push(Tag::Int64_8.into());
+                data.extend_from_slice(&i.to_le_bytes()[0..1]);
+            } else if mbs <= 16 {
+                data.push(Tag::Int64_16.into());
+                data.extend_from_slice(&i.to_le_bytes()[0..2]);
+            } else if mbs <= 24 {
+                data.push(Tag::Int64_24.into());
+                data.extend_from_slice(&i.to_le_bytes()[0..3]);
+            } else if mbs <= 32 {
+                data.push(Tag::Int64_32.into());
+                data.extend_from_slice(&i.to_le_bytes()[0..4]);
+            } else if mbs <= 40 {
+                data.push(Tag::Int64_40.into());
+                data.extend_from_slice(&i.to_le_bytes()[0..5]);
+            } else if mbs <= 48 {
+                data.push(Tag::Int64_48.into());
+                data.extend_from_slice(&i.to_le_bytes()[0..6]);
+            } else if mbs <= 56 {
+                data.push(Tag::Int64_56.into());
+                data.extend_from_slice(&i.to_le_bytes()[0..7]);
+            } else {
+                data.push(Tag::Int64.into());
+                data.extend_from_slice(&i.to_le_bytes());
+            }
```

Comment:
can we just use https://docs.rs/integer-encoding/latest/integer_encoding/trait.VarInt.html?

Reply:
I think that will be less efficient than what we're doing here, which we're able to do because the tag gives us room to encode the length for free.

Reply:
ah yes, nice!